File: APKInstaller/Installer.vb
﻿Imports System.IO
Imports System.Threading
Imports ICSharpCode.SharpZipLib.Zip

Public Class Installer
    Private Delegate Sub deviceIdCallback(ByVal deviceId As String)
    Private GUI As Main
    Private lblStatus As MaterialSkin.Controls.MaterialLabel
    Private txtUserInput As MaterialSkin.Controls.MaterialSingleLineTextField
    Private stopRequested As Boolean = False
    Private update As Boolean = True
    Private showCompletionMessage As Boolean = True
    Private adbCache As String
    Private aaptLocation As String
    Private force As Boolean

    Sub New(ByRef entry As Main, ByRef statusLabel As MaterialSkin.Controls.MaterialLabel, ByRef userInputTextbox As MaterialSkin.Controls.MaterialSingleLineTextField)
        GUI = entry
        lblStatus = statusLabel
        txtUserInput = userInputTextbox
    End Sub

    Private Function GetShell32Folder(folderPath As String) As Shell32.Folder
        Dim shellAppType As Type = Type.GetTypeFromProgID("Shell.Application")
        Dim Shell = Activator.CreateInstance(shellAppType)
        Return CType(shellAppType.InvokeMember("NameSpace", Reflection.BindingFlags.InvokeMethod, Nothing, Shell, New Object() {folderPath}), Shell32.Folder)
    End Function

    Private Sub WaitForDevice(adbLocation As String)
        Dim adbWait As New Process()
        adbWait.StartInfo.FileName = adbLocation
        adbWait.StartInfo.Arguments = "wait-for-device"
        adbWait.StartInfo.UseShellExecute = False
        adbWait.StartInfo.CreateNoWindow = True
        GUI.SetText(lblStatus, "Waiting for device.")
        adbWait.Start()

        Dim caret As Integer = 0
        While Not adbWait.HasExited
            GUI.UseWaitCursor = True
            GUI.ShowProgressAnimation(True, False)
            If caret = 0 Then
                GUI.SetText(lblStatus, "Waiting for device.")
            ElseIf caret = 1 Then
                GUI.SetText(lblStatus, "Waiting for device..")
            Else
                GUI.SetText(lblStatus, "Waiting for device...")
                caret = -1
            End If
            caret += 1

            Thread.Sleep(1000)
            Threading.Thread.Yield()
            If stopRequested Then
                MsgBox("The install has been aborted", CType(MsgBoxStyle.OkOnly + MsgBoxStyle.Exclamation, MsgBoxStyle), "APK Install Aborted")
                Exit Sub
            End If
        End While
        GUI.ShowProgressAnimation(False, False)
        adbWait.Close()
    End Sub

    Function GetAdbExecutable() As String
        If adbCache IsNot Nothing And (adbCache Is "adb" Or File.Exists(adbCache)) Then
            Return adbCache
        End If

        Dim adb As New Process
        adb.StartInfo.Arguments = "version"
        adb.StartInfo.FileName = "adb"
        adb.StartInfo.CreateNoWindow = True
        adb.StartInfo.UseShellExecute = False
        Try
            adb.Start()
            adb.WaitForExit()
            If (adb.ExitCode = 0) Then
                adbCache = "adb"
                Return "adb"
            End If
        Catch ex As Exception
            ' Continue on
        End Try

        ' Fall-back to app version
        Dim tempFileName As String = Path.GetTempFileName()
        File.Delete(tempFileName)
        Directory.CreateDirectory(tempFileName)
        Dim platformToolsZip As String = tempFileName + "\platform-tools.zip"
        File.WriteAllBytes(platformToolsZip, My.Resources.platform_tools_r23_1_0_windows)
        Dim androidPlatformTools As String = tempFileName + "\platform-tools"
        UnzipFromStream(New FileStream(platformToolsZip, FileMode.Open), androidPlatformTools)
        adbCache = androidPlatformTools & "\platform-tools\adb.exe"
        Return adbCache
    End Function



    Public Sub UnzipFromStream(zipStream As FileStream, outFolder As String)
        Dim zipInputStream = New ZipInputStream(zipStream)
        Dim zipEntry = zipInputStream.GetNextEntry()
        Dim buffer(4096) As Byte    ' 4K Is optimum
        While (zipEntry IsNot Nothing)
            Dim entryFileName = zipEntry.Name
            ' Convert UNIX paths to the current platform path
            entryFileName = entryFileName.Replace("/", Path.DirectorySeparatorChar)
            Dim fullZipToPath = Path.Combine(outFolder, entryFileName)
            Dim directoryName = Path.GetDirectoryName(fullZipToPath)

            If (directoryName.Length > 0) Then
                Directory.CreateDirectory(directoryName)
            End If
            If Not (directoryName & "\").Equals(fullZipToPath) Then
                Using _streamWriter As FileStream = File.Create(fullZipToPath)
                    ICSharpCode.SharpZipLib.Core.StreamUtils.Copy(zipInputStream, _streamWriter, buffer)
                End Using
            End If


            zipEntry = zipInputStream.GetNextEntry()
        End While
    End Sub

    Private Sub GetDeviceId(callback As deviceIdCallback)
        Dim deviceChooser As New dlgDeviceChoose()
        Dim result = deviceChooser.getUserInput(GetAdbExecutable)
        If result = DialogResult.OK Then
            If (MsgBox("Is the device """ + deviceChooser.getDevice + """ correct?", CType(MsgBoxStyle.YesNo + MsgBoxStyle.Question, MsgBoxStyle)) = MsgBoxResult.Yes) Then
                callback(deviceChooser.getDevice)
            Else
                GetDeviceId(callback)
            End If
        ElseIf result = DialogResult.Cancel Then
            abort()
        Else
            callback(deviceChooser.getDevice)
        End If
    End Sub

    Sub AddFilesToInstall(files() As String)
        Dim location As String = ""
        If GetFilesToInstall.Length > 0 And GetFilesToInstall(0) IsNot "" Then
            If MsgBox("There are other APK files to install. Do you want to keep them and add this to install?", CType(MsgBoxStyle.Question + MsgBoxStyle.YesNo, MsgBoxStyle)) = MsgBoxResult.Yes Then
                location = txtUserInput.Text & ";"
            End If
        End If

        For Each path As String In files
            If path Is Nothing Then
                Continue For
            End If

            Dim invalidPackage = Not path.ToLower.EndsWith(".apk")

            Try
                If (Not invalidPackage And packageName(path) Is "") Then
                    invalidPackage = True
                End If
            Catch ex As Exception
                MsgBox(ex.Message)
            End Try

            If Not invalidPackage Then
                location += path + ";"
            Else
                MsgBox("""" & path & """" & " is not a valid Android app. Please verify that the file ends with "".APK"".", CType(MsgBoxStyle.OkOnly + MsgBoxStyle.Exclamation, MsgBoxStyle), "Invalid File")
            End If
        Next

        If location.EndsWith(";") Then
            location = location.Substring(0, location.Length - 1)
        End If
        txtUserInput.Text = location
    End Sub

    Function VerifyFilesToInstall() As Boolean
        Dim apkFiles = GetFilesToInstall()
        If apkFiles.Length = 0 Then
            Return False
        End If

        For Each apkfile As String In apkFiles
            Dim exists = File.Exists(apkfile)
            If Not exists Then
                Return False
            End If
        Next

        Return True
    End Function

    Sub StartInstall()
        Dim thread As New Thread(New ThreadStart(AddressOf Install))
        thread.Start()
    End Sub

    Sub ConfigureReInstall(enabled As Boolean)
        update = enabled
    End Sub

    Sub ConfigureForce(enabled As Boolean)
        force = enabled
    End Sub

    Sub ShowCompletionMessageWhenFinished(show As Boolean)
        showCompletionMessage = show
    End Sub

    Private Sub Install()
        Dim adbLocation As String = GetAdbExecutable()

        ' Wait until an Android device is connected
        WaitForDevice(adbLocation)
        If stopRequested Then
            Return
        End If

        ' Device Found, check if multiple devices are connected, and figure out the device to install to
        Dim deviceId = ""
        GUI.SetText(lblStatus, "Checking device(s)...This may be a moment or two")
        GUI.ShowProgressAnimation(True, False)
        GetDeviceId(Sub(ByVal _DeviceId As String)
                        deviceId = _DeviceId
                    End Sub)
        While deviceId Is ""
            If (stopRequested) Then
                GUI.SetText(lblStatus, "The install did not complete successfully :(")
                GUI.ShowProgressAnimation(False, False)
                Return

            End If
            Thread.Sleep(50)
        End While
        GUI.ShowProgressAnimation(False, False)

        GUI.SetText(lblStatus, "Starting Installs...")
        Dim filesToInstall As String() = GetFilesToInstall()
        'pgbStatus.Step = 100 / filesToInstall.Length
        Dim installStatus As String = ""
        Dim installAborted = False
        GUI.ShowProgressAnimation(True, True, CInt(100 / filesToInstall.Length))

        For Each file As String In filesToInstall
            While True
                If (force) Then
                    ForceRemovePackage(file)
                End If

                Dim adb As New Process()
                Dim arguments As String = "-s " & deviceId & " install "
                If update Then
                    arguments += "-r "
                End If
                adb.StartInfo.Arguments = arguments & """" & file & """"
                adb.StartInfo.CreateNoWindow = True
                adb.StartInfo.UseShellExecute = False
                adb.StartInfo.FileName = adbLocation
                adb.StartInfo.RedirectStandardOutput = True
                installStatus += adb.StartInfo.FileName + " " + adb.StartInfo.Arguments
                installStatus += vbCrLf
                GUI.SetText(GUI.lblStatus, installStatus)
                adb.Start()
                While Not adb.HasExited
                    If stopRequested Then
                        MsgBox("The install has been aborted!", CType(MsgBoxStyle.OkOnly + MsgBoxStyle.Exclamation, MsgBoxStyle), "APK Install Aborted")
                        Return
                    End If
                End While
                installStatus += adb.StandardOutput.ReadToEnd
                GUI.SetText(GUI.lblStatus, installStatus)

                If Not adb.ExitCode = 0 Then
                    installStatus += adb.StandardOutput.ReadToEnd
                    GUI.SetText(GUI.lblStatus, installStatus)
                    Dim result As MsgBoxResult = MsgBox("The installation of """ & file & """ did not succeed." & vbCrLf & "Details: ADB exited with error code " & adb.ExitCode, CType(MsgBoxStyle.Exclamation + MsgBoxStyle.AbortRetryIgnore, MsgBoxStyle))
                    If result = MsgBoxResult.Abort Then
                        Exit For
                    ElseIf result = MsgBoxResult.Retry Then
                        Continue While
                    End If
                Else
                    GUI.StepProgressBar()
                End If
                Exit While
            End While
        Next
        GUI.UseWaitCursor = False
        If Not installAborted Then
            GUI.SetText(lblStatus, "Done!")
        Else
            GUI.SetText(lblStatus, "Failure! Details: " & vbCrLf & installStatus)
        End If

        If Not installAborted And showCompletionMessage Then
            MsgBox("The installation of the APK(s) has finished successfully!", CType(MsgBoxStyle.Information + MsgBoxStyle.OkOnly, MsgBoxStyle), "APK Install Finished")
        End If
    End Sub

    Function GetFilesToInstall() As String()
        Return txtUserInput.Text.Split(CType(";", Char()))
    End Function

    Sub abort()
        stopRequested = True
    End Sub

    Function ForceRemovePackage(ByVal apkFile As String) As Boolean
        Dim adb As New Process()
        adb.StartInfo.FileName = GetAdbExecutable()
        adb.StartInfo.Arguments = "uninstall " & packageName(apkFile)
        adb.StartInfo.CreateNoWindow = True
        adb.StartInfo.UseShellExecute = False
        adb.Start()
        adb.WaitForExit()

        Return adb.ExitCode = 0
    End Function

    Function getAaptLocation() As String
        If (aaptLocation IsNot Nothing) Then
            If (File.Exists(aaptLocation)) Then
                Return aaptLocation
            End If
        End If
        Dim aapt = ""
        While aapt Is ""
            Try
                Dim temp = Path.GetTempFileName()
                File.Delete(temp)
                Directory.CreateDirectory(temp)
                aapt = Path.Combine(temp, "aapt.exe")
            Catch ex As Exception

            End Try
        End While

        File.WriteAllBytes(aapt, My.Resources.aapt_23_0_3_win)
        If File.Exists(aapt) Then
            aaptLocation = aapt
            Return aapt
        Else
            Throw New IOException("Failed to build aapt.exe")
        End If
    End Function

    Function packageName(ByVal apkFile As String) As String
        Dim aapt As New Process()
        aapt.StartInfo.FileName = getAaptLocation()
        aapt.StartInfo.Arguments = "dump badging """ & apkFile & """"
        aapt.StartInfo.CreateNoWindow = True
        aapt.StartInfo.UseShellExecute = False
        aapt.StartInfo.RedirectStandardOutput = True
        aapt.Start()

        Const parseFor = "package: name="
        Dim package = ""
        Dim line As String = aapt.StandardOutput.ReadLine
        Try
            While line IsNot Nothing
                'Detect interrupts
                Threading.Thread.Sleep(1)

                line = line.Trim
                If aapt.HasExited Then
                    If Not aapt.ExitCode = 0 Then
                        Throw New Exception("AAPT Failed. Exit: " & aapt.ExitCode)
                    End If
                End If
                If line.Contains(parseFor) Then
                    Dim versionName As String = line.Substring(line.IndexOf(parseFor) + parseFor.Length)
                    package = versionName.Substring(versionName.IndexOf("'") + 1)
                    package = package.Substring(0, package.IndexOf("'"))
                    Exit While

                End If
                line = aapt.StandardOutput.ReadLine
            End While
            If aapt.HasExited Then
                If Not aapt.ExitCode = 0 Then
                    Throw New Exception("AAPT Failed. Exit: " & aapt.ExitCode)
                End If
            End If
        Catch ex As Exception
            Dim exception As New IOException("Failed to acquire package name", ex)
            Throw exception
        End Try
        Return package
    End Function
End Class

File: AppInstaller/Installer.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Threading;
using System.Windows.Forms;
using APKInstaller.i18n;
using MaterialSkin.Controls;

namespace APKInstaller
{
    /// <summary>
    ///     Handler of installing Android packages
    /// </summary>
    public class Installer
    {
        readonly LinkedList<string> _filesToInstall = new LinkedList<string>();
        readonly Main _gui;
        readonly MaterialSingleLineTextField _txtUserInput;
        Label _lblStatus;
        bool _stopRequested;

        /// <summary>
        ///     Creates a new installer instance
        /// </summary>
        /// <param name="entry">The main GUI of the application</param>
        /// <param name="statusLabel">The label to use as installer statuses</param>
        /// <param name="userInputTextBox">The user input textbox</param>
        [SuppressMessage("Microsoft.Design", "CA1045:DoNotPassTypesByReference", MessageId = "2#")]
        [SuppressMessage("Microsoft.Design", "CA1045:DoNotPassTypesByReference", MessageId = "1#")]
        [SuppressMessage("Microsoft.Design", "CA1045:DoNotPassTypesByReference", MessageId = "0#")]
        public Installer(ref Main entry, ref Label statusLabel, ref MaterialSingleLineTextField userInputTextBox)
        {
            _gui = entry;
            _lblStatus = statusLabel;
            _txtUserInput = userInputTextBox;
        }

        /// <summary>
        ///     Determines if the operations performed are update (reinstall) operations, or not
        /// </summary>
        /// <returns>true if the packages are going to be reinstalled; otherwise false</returns>
        public bool Reinstall { get; set; } = true;

        /// <summary>
        ///     Whether to force the package installation or not
        /// </summary>
        /// <returns>true if the packages will be forcibly installed</returns>
        public bool Force { get; set; }

        /// <summary>
        ///     Shows a completion message after the installation of all packages is finished
        /// </summary>
        /// <returns>if a completion message will be shown after the installation job is finished</returns>
        public bool CompletionMessageWhenFinished { get; set; }

        public bool UseMultiFileDialog { get; set; }

        public string FilesToInstallDescription { get; private set; } = "";

        public void AddFilesToInstall(string[] files)
        {
            AddFilesToInstall(files, false);
        }

        public void AddFilesToInstall(string[] files, bool clear)
        {
            AddFilesToInstall(files, clear, true);
        }

        /// <summary>
        ///     Adds files to be installed
        /// </summary>
        /// <param name="files">files to be installed</param>
        /// <param name="clear"></param>
        /// <param name="notifyUser"></param>
        //[Log("App Installer Debug")]
        public void AddFilesToInstall(string[] files, bool clear, bool notifyUser)
        {
            if (files == null)
                throw new ArgumentNullException(nameof(files));

            if (clear)
                _filesToInstall.Clear();

            foreach (var path in files)
            {
                // Prevent bugs in detection routine
                if (path == null)
                    continue;

                var path2 = path.Trim();

                // Determine if the file is already in the installer list, if so skip it
                if (_filesToInstall.Contains(path2))
                    continue;

                try
                {
                    if (ValidateFile(path, notifyUser))
                        _filesToInstall.AddLast(path);
                }
                catch (IOException ex)
                {
                    MessageBox.Show(ex.Message, UIStrings.Invalid_File, MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                }
            }

            _txtUserInput.Text = GenerateFileListSummary();
            FilesToInstallDescription = _txtUserInput.Text;
        }

        //[Log("App Installer Debug")]
        public void RemoveFile(string path)
        {
            _filesToInstall.Remove(path);
        }

        public static bool ValidateFile(string path) => ValidateFile(path, false);

        /// <summary>
        ///     Determines if the given file is a valid APK file
        /// </summary>
        /// <param name="path">the file location for the APK file</param>
        /// <param name="notifyUser">true will notify the user of the problem via dialogs, false will not</param>
        /// <returns>true if the file is a valid APK, false otherwise</returns>
        //[Log("App Installer Debug")]
        public static bool ValidateFile(string path, bool notifyUser)
        {
            if (path == null)
                throw new ArgumentNullException(nameof(path));

            // Check to see if the file exists
            if (!File.Exists(path))
            {
                if (notifyUser)
                    MessageBox.Show(
                        Directory.Exists(path) ? "\"" + path + "\" " + UIStrings.isDirError : UIStrings.fileDoesNotExist,
                        UIStrings.Invalid_File, MessageBoxButtons.OK, MessageBoxIcon.Warning);
                return false;
            }

            try
            {
                // Check for the correct extension and that the file can be correctly parsed as an APK
                if (path.ToUpper(CultureInfo.CurrentCulture)
                        .EndsWith(".APK", StringComparison.CurrentCultureIgnoreCase) &
                    (AndroidTools.PackageName(path) != "")) return true;
                if (notifyUser)
                    MessageBox.Show($@"""{path}"" {UIStrings.invalidApk}", UIStrings.Invalid_File, MessageBoxButtons.OK,
                        MessageBoxIcon.Warning);
                return false;
            }
            catch (IOException ex)
            {
                if (notifyUser)
                    MessageBox.Show($@"""{path}"" {UIStrings.invalidApk}", UIStrings.Invalid_File, MessageBoxButtons.OK,
                        MessageBoxIcon.Warning);
                return false;
            }
        }

        string GenerateFileListSummary()
        {
            switch (_filesToInstall.Count)
            {
                case 0:
                    UseMultiFileDialog = false;
                    return "";
                case 1:
                    UseMultiFileDialog = false;
                    return _filesToInstall.First.Value;
            }
            UseMultiFileDialog = true;
            return _filesToInstall.ElementAt(0) + ", " + _filesToInstall.ElementAt(1) +
                   (_filesToInstall.Count > 2 ? "..." : "");
        }

        /// <summary>
        ///     Verifies that all of the files to be installed are valid
        /// </summary>
        /// <returns></returns>
        public bool VerifyFilesToInstall(bool excuseNoFiles = false)
        {
            dynamic apkFiles = GetFilesToInstall();

            // Nothing to install
            if (apkFiles.Length == 0)
                return excuseNoFiles;

            foreach (var apkFile in apkFiles)
                if (!ValidateFile(apkFile))
                    return false;
            return true;
        }

        /// <summary>
        ///     Starts the install of the APK files
        /// </summary>
        public void StartInstall()
        {
            var thread = new Thread(Install);
            thread.Start();
        }

        /// <summary>
        ///     Installs the APK packages
        /// </summary>
        void Install()
        {
            // Wait until an Android device is connected
            dynamic waitForDeviceReturn = WaitForDevice();
            if (waitForDeviceReturn != ErrorCode.Success)
            {
                HandleWaitForDeviceError(waitForDeviceReturn);
                return;
            }


            // Device Found, check if multiple devices are connected, and figure out the device to install to
            dynamic deviceId = AcquireDeviceId();
            if (deviceId == null)
            {
                _gui.ResetGui(UIStrings.installAborted);
                return;
            }

            // Begin device install sections
            _gui.SetText(ref _lblStatus, UIStrings.startingInstalls);
            var filesToInstall = GetFilesToInstall();
            dynamic installStatus = "";
            dynamic installAborted = false;
            dynamic installHasFailed = false;
            _gui.ShowProgressAnimation(true, true, Convert.ToInt32(100 / filesToInstall.Length));
            const int maxRetryCount = 3;
            foreach (var file in filesToInstall)
            {
                // Check if the install has been aborted, if so stop the installs
                if (installAborted)
                {
                    break;
                }

                // Retry Loop
                dynamic packageInstallSuccess = false;
                for (var retry = 0; (retry <= maxRetryCount) && !packageInstallSuccess; retry++)
                {
                    //Display a status
                    _gui.SetText(ref _gui.lblStatus,
                        retry == 0
                            ? UIStrings.installing
                            : UIStrings.retrying_install + " \"" + file + UIStrings.onto_the_device + deviceId + "\"");
                    ErrorCode installAttempt = PackageInstallAttempt(deviceId, file);
                    switch (installAttempt)
                    {
                        case ErrorCode.Abort:
                            // ReSharper disable once RedundantAssignment
                            installAborted = true;
                            return;
                        case ErrorCode.Failure1:
                        case ErrorCode.Failure2:
                            continue;
                        case ErrorCode.Success:
                        case ErrorCode.Ignore:
                            packageInstallSuccess = true;
                            break;
                    }
                }
                // End retry loop
                if (packageInstallSuccess) continue;
                MessageBox.Show($@"The file ""{file}"" couldn't be installed.", UIStrings.Installer_Install_Failed_install,
                    MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                installHasFailed = true;
            }
            // for each file
            _gui.UseWaitCursor = false;
            if (installAborted)
            {
                if (CompletionMessageWhenFinished)
                {
                    _gui.ResetGui(UIStrings.failure + UIStrings.details + "\n" + installStatus);
                }
                else
                {
                    MessageBox.Show(UIStrings.failure + UIStrings.details + "\n" + installStatus, "Install Aborted",
                        MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                    _gui.Close();
                }
            }
            else
            {
                if (CompletionMessageWhenFinished)
                {
                    dynamic message = UIStrings.installationOf +
                                      (filesToInstall.Length == 1 ? UIStrings.apk : UIStrings.apks);
                    if (installHasFailed)
                        MessageBox.Show(message + UIStrings.hasFinishedError);
                    else
                        MessageBox.Show(message + UIStrings.hasFinishedSucess, UIStrings.apkInstallFinished,
                            MessageBoxButtons.OK, MessageBoxIcon.Information);
                    _gui.ResetGui(UIStrings.done);
                }
                else
                {
                    _gui.SetText(ref _gui.lblStatus, UIStrings.done);
                }
            }
            //FinishInstall(If(installAborted, UIStrings.failure & UIStrings.details & vbCrLf & installStatus, UIStrings.done))
            // End device install section
        }

        ErrorCode PackageInstallAttempt(string deviceId, string file)
        {
            if (Force)
                ForceRemovePackage(file);

            dynamic adbCode = InstallSinglePackage(deviceId, file);
            // Install Single Package failed catastrophically when this happened
            if (adbCode == null)
                return ErrorCode.Failure1;

            int result;
            // Checks if the result is a return code
            if (!int.TryParse(adbCode, out result)) return HandleInstallFailure(file, adbCode);

            if (result != 0) return HandleInstallFailure(file, result);
            _gui.StepProgressBar();
            return ErrorCode.Success;
            // the result is an details error message
        }

        void HandleWaitForDeviceError(ErrorCode waitForDeviceReturn)
        {
            dynamic message = "An unknown error while waiting for the device";
            switch (waitForDeviceReturn)
            {
                case ErrorCode.FailureTimeout:
                    message = $"{UIStrings.timeoutWaiting}\n" +
                              $"{UIStrings.userTroubleshootingA1}\n" +
                              $"{UIStrings.userTroubleshootingA2}";
                    break;
                case ErrorCode.Abort:
                    message = UIStrings.installAborted;
                    break;
                case ErrorCode.Success:
                    break;
                case ErrorCode.Ignore:
                    break;
                case ErrorCode.Failure1:
                    break;
                case ErrorCode.Failure2:
                    break;
                case ErrorCode.Failure3:
                    break;
                default:
                    throw new ArgumentOutOfRangeException(nameof(waitForDeviceReturn), waitForDeviceReturn, null);
            }

            if (CompletionMessageWhenFinished)
            {
                _gui.ResetGui(message);
            }
            else
            {
                MessageBox.Show(message, UIStrings.Device_Error, MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                _gui.Close();
            }
        }

        ErrorCode WaitForDevice()
        {
            dynamic counter = 0;
            using (var adbWait = AndroidTools.RunAdb("wait-for-device", false, true, false))
            {
                var caret = 0;
                _gui.UseWaitCursor = true;
                _gui.ShowProgressAnimation(true, false);
                while (!adbWait.HasExited)
                {
                    string currentMessage = null;
                    switch (caret)
                    {
                        case 0:
                            currentMessage = UIStrings.waitForDevice1;
                            break;
                        case 1:
                            currentMessage = UIStrings.waitForDevice2;
                            break;
                        default: // caret == 2
                            currentMessage = UIStrings.waitForDevice3;
                            caret = -1;
                            break;
                    }
                    caret++;

                    // Display help message after 10 seconds
                    if (counter > 10)
                        currentMessage += $"\n{UIStrings.userTroubleshootingA1}\n{UIStrings.userTroubleshootingA2}";
                    _gui.SetText(ref _lblStatus, currentMessage);

                    Thread.Sleep(1000);
                    Thread.Yield();
                    if (_stopRequested)
                    {
                        MessageBox.Show(UIStrings.installAborted, UIStrings.APK_Install_Aborted, MessageBoxButtons.OK,
                            MessageBoxIcon.Exclamation);
                        return ErrorCode.Abort;
                    }
                    counter++;
                
                    if (counter > 60) // Timeout after 60 seconds
                        return ErrorCode.FailureTimeout;
                }
                _gui.ShowProgressAnimation(false, false);
            }
            return 0;
        }

        string AcquireDeviceId()
        {
            _gui.SetText(ref _lblStatus, UIStrings.checkDevices);
            _gui.ShowProgressAnimation(true, false);

            string deviceId = null;
            // ReSharper disable once RedundantAssignment
            GetDeviceId(id => { deviceId = id; });
            while (deviceId == null)
            {
                if (_stopRequested)
                {
                    _gui.SetText(ref _lblStatus, UIStrings.installAborted);
                    _gui.ShowProgressAnimation(false, false);
                    return null;
                }
                Thread.Sleep(50);
            }

            _gui.ShowProgressAnimation(false, false);

            return deviceId;
        }

        void GetDeviceId(DeviceIdCallback callback)
        {
            while (true)
            {
                var deviceChooser = new DeviceChooserDialog();
                dynamic result = deviceChooser.GetUserInput();
                if ((result == DialogResult.OK) | (result == null))
                    if (
                        MessageBox.Show(UIStrings.correctDevice1 + deviceChooser.Device + UIStrings.correctDevice2,
                            UIStrings.Verify_Device, MessageBoxButtons.YesNo, MessageBoxIcon.Question) == DialogResult.Yes)
                        callback(deviceChooser.Device);
                    else
                        continue;
                else if (result == DialogResult.Cancel)
                    Abort();
                else
                    callback(deviceChooser.Device);
                //End Using
                break;
            }
        }

        public string[] GetFilesToInstall()
        {
            return _filesToInstall.ToArray();
        }

        public static bool ForceRemovePackage(string apkFile)
        {
            using (var adb = AndroidTools.RunAdb("uninstall " + AndroidTools.PackageName(apkFile), false, true, true))
            {
                return adb.ExitCode == 0;
            }
        }

        string InstallSinglePackage(string deviceId, string file)
        {
            const string abortKey = "ABORT";

            using (
                var adb =
                    AndroidTools.RunAdb($@"-s {deviceId} install {(Reinstall ? "-r " : "")}""{file}""",
                        true, true, false))
            {
                dynamic adbStandardOut = adb.StandardOutput;
                dynamic haltInstall = false;

                // Process aborts while the adb process is running
                // Check for and identify failures during the adb install
                dynamic currentLine = adbStandardOut.ReadLine();
                while (!adb.HasExited | (currentLine != null))
                {
                    // user requested stop
                    if (_stopRequested)
                    {
                        if (
                            MessageBox.Show(UIStrings.abortCurrentApk, UIStrings.APK_Install_Aborted, MessageBoxButtons.YesNo,
                                MessageBoxIcon.Question) == DialogResult.Yes)
                        {
                            try
                            {
                                if (adb.HasExited)
                                    adb.Kill();
                            }
                            catch (InvalidOperationException ex)
                            {
                            }
                            MessageBox.Show(UIStrings.installAborted, UIStrings.APK_Install_Aborted, MessageBoxButtons.OK,
                                MessageBoxIcon.Exclamation);
                            return abortKey;
                        }
                        haltInstall = true;
                        _stopRequested = false;
                    }


                    while ((currentLine != null) && !_stopRequested)
                    {
                        if (currentLine.StartsWith("Failure", StringComparison.OrdinalIgnoreCase))
                        {
                            dynamic startPos = currentLine.IndexOf("[", StringComparison.OrdinalIgnoreCase);
                            dynamic endPos = currentLine.IndexOf("]", StringComparison.OrdinalIgnoreCase);
                            dynamic errorMessage = "Unknown Failure";
                            if ((startPos < endPos) & (startPos != -1) & (endPos != -1))
                                errorMessage = currentLine.Substring(startPos + 1, endPos - startPos - 1);
                            return errorMessage;
                        }
                        currentLine = adbStandardOut.ReadLine();
                    }
                }

                // Display an abort message
                if (!haltInstall) return adb.ExitCode.ToString();

                MessageBox.Show(UIStrings.installAborted, UIStrings.installAborted, MessageBoxButtons.OK,
                    MessageBoxIcon.Exclamation);
                return abortKey;
            }
        }

        ErrorCode HandleInstallFailure(string file, int code)
        {
            return HandleInstallFailure(file, UIStrings.adbExitErrorCode + code);
        }

        ErrorCode HandleInstallFailure(string file, string message)
        {
            var result =
                MessageBox.Show(
                    UIStrings.unsuccessfulInstall1 + file + UIStrings.unsuccessfulInstall2 + "\n" + UIStrings.details +
                    message, "Install Failed", MessageBoxButtons.AbortRetryIgnore, MessageBoxIcon.Question);
            switch (result)
            {
                case DialogResult.Abort:
                    return ErrorCode.Abort;
                case DialogResult.Ignore:
                    // Don't retry
                    return ErrorCode.Ignore;
                default:
                    for (var second = 5; second >= 1; second += -1)
                    {
                        dynamic retryMessage = UIStrings.retryIn + second +
                                               (second == 1 ? UIStrings.second : UIStrings.seconds);
                        _gui.SetText(ref _gui.lblStatus, retryMessage);
                    }

                    return ErrorCode.Failure2;
            }
        }

        /// <summary>
        ///     Aborts the install
        /// </summary>
        public void Abort()
        {
            _stopRequested = true;
        }

        public void Reset()
        {
            _stopRequested = false;
        }

        delegate void DeviceIdCallback(string deviceId);
    }
}